File: src/models/tree.js
const Node = require('./node')
class Tree {
  constructor () {
    this.rootNode = new Node(null, 'root', 0)
  }
  addWord(word) {
    if (word && word.length) {
      let currentNode = this.rootNode
      for (let i = 0; i < word.length; i++) {
        currentNode = currentNode.addChild(word[i])
      }
      currentNode.isEndOfWord = true
    }
  }
  addWords(words) {
    for (let i = 0; i < words.length; i++) {
      this.addWord(words[i])
    }
  }
  isConcatenated(word, isOriginalWord = true) {
    for (let index = 0, currentNode = this.rootNode, nextNode = null; index < word.length; index++, currentNode = nextNode) {
      nextNode = currentNode.childNodes[word[index]]
      if (!nextNode) return false
      if (nextNode.depth === word.length) {
        if (isOriginalWord) return false
        if (nextNode.isEndOfWord) return true
        return false
      }
      if (nextNode.isEndOfWord && this.isConcatenated(word.substring(nextNode.depth), false)) return true
    }
    return false
  }
}
module.exports = Tree

File: py_src/classes/tree.py

from node import Node

class Tree:

    def __init__(self):
        self.root_node = Node('root', 0, {})

    def add_word(self, word):
        if word:
            current_node = self.root_node
            for char in word:
                current_node = current_node.add_child(char)
            current_node.is_end_of_word = True

    def add_words(self, words):
        for word in words:
            self.add_word(word)

    def print_tree(self):
        self.root_node.print_node()

    def is_concatenated(self, word, is_original_word = True):
        current_node = self.root_node
        for char in word:
            try:
                next_node = current_node.child_nodes[char]
            except:
                return False
            if next_node.depth == len(word):
                if is_original_word: return False
                if next_node.is_end_of_word: return True
                return False
            if next_node.is_end_of_word and \
                    self.is_concatenated(word[next_node.depth:], False):
                return True
            current_node = next_node
        return False
